### ISSUE TITLE
Cannot use derived Deserialize impl for enums

### ISSUE THREAD

**Issue by Xion** [samscott89/serde_qs#6]
When trying to deserialize a query string containing enum values encoded as strings, it fails with an "enum expected" error message.

Here's a repro code (with deserialization via serde_json thrown in for comparison):

```rust
             extern crate serde;
#[macro_use] extern crate serde_derive;
             extern crate serde_json;
             extern crate serde_qs;


#[derive(Deserialize, Debug)]
#[serde(rename_all = "lowercase")]
enum TestEnum { Foo, Bar }

#[derive(Deserialize, Debug)]
struct TestStruct {
    e: TestEnum,
}


fn main() {
    println!("serde_json: {:?}", serde_json::from_str::<TestStruct>(r#"{"e": "foo"}"#));
    println!("serde_qs: {:?}", serde_qs::from_str::<TestStruct>("e=foo"));
}
```

This produces:
```
serde_json: Ok(TestStruct { e: Foo })
serde_qs: Err(Error { err: "invalid type: string \"foo\", expected enum TestEnum" })
```

If we don't use `#[derive(Deserialize)]` however, and implement it manually for the enum type (like [here](https://github.com/Xion/rofld/blob/4ec9d11295ce51130d3982045e1b5cde4855ec9e/src/model/de/alignment.rs#L33)), the enums will be correctly deserialized.

**Comment by samscott89** [comment]
Hey,
Currently we only support tagged enums (see the example: https://github.com/samscott89/serde_qs/blob/master/examples/introduction.rs#L135).

So you could have the following:
```rust
#[derive(Deserialize, Debug)]
#[serde(rename_all = "lowercase")]
#[serde(tag = "type", content = "value")]
enum TestEnum { Foo, Bar }

#[derive(Deserialize, Debug)]
struct TestStruct {
    e: TestEnum,
}

fn main() {
    println!("serde_qs: {:?}", serde_qs::from_str::<TestStruct>("e[type]=foo"));
}
```

I will revisit the logic though, because it should at least be possible to deserialize UnitVariants as in your example.

**Comment by Xion** [comment]
Yes, that would be great to have :) I'll just continue using the manual impl workaround for now. Thanks!

**Comment by samscott89** [comment]
Okay, I've finished the overdue refactor of the code. It should now happily serialize/deserialize all forms of enums in the following way.

Suppose you have a struct and enum like so:
```rust

    #[derive(Deserialize, Debug, PartialEq)]
    enum E {
        B,
        S(String),
    }

    #[derive(Deserialize, Debug, PartialEq)]
    struct Query {
        e: E,
    }
```

Then this would be de/serialized from/as either `"e=B"` or `"e[S]=Hello+World"` depending on which variant you have.

**Comment by samscott89** [comment]
Let me know if this works for you and I'll publish this is `0.2`.

**Comment by Xion** [comment]
Almost :) One problem I encountered is that since tuple_struct is no longer `forward_to_deserialize_any`, it now seems to mandate always being deserialized from a sequence. This breaks if a tuple struct is using a custom deserializer.

Here's a (somewhat complex) repro case:
```rust
             extern crate serde;
#[macro_use] extern crate serde_derive;
#[macro_use] extern crate serde_json;
             extern crate serde_qs;


use std::fmt;

use serde::de::{self, Deserialize, Visitor};


#[derive(Deserialize, Debug, Clone, Copy)]
#[serde(rename_all = "lowercase")]
enum TestEnum { Foo, Bar }

#[derive(Deserialize, Debug)]
struct TestStruct {
    e: TestEnum,
}


#[derive(Debug)]
struct TupleStruct(TestEnum, TestStruct);

impl<'de> Deserialize<'de> for TupleStruct {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
        where D: de::Deserializer<'de>
    {
        deserializer.deserialize_tuple_struct(
            "TupleStruct", 2, TupleStructVisitor)
    }
}

struct TupleStructVisitor;
impl<'de> Visitor<'de> for TupleStructVisitor {
    type Value = TupleStruct;

    fn expecting(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "TupleStruct representation")
    }

    fn visit_map<V>(self, map: V) -> Result<Self::Value, V::Error>
        where V: de::MapAccess<'de>
    {
        let inner_de = de::value::MapAccessDeserializer::new(map);
        let test_struct: TestStruct = Deserialize::deserialize(inner_de)?;

        let result = TupleStruct(test_struct.e, test_struct);
        Ok(result)
    }
}


#[derive(Deserialize, Debug)]
struct NestedStruct {
    ts: Vec<TupleStruct>,
}


fn main() {
    println!("serde_json: {:?}", serde_json::from_str::<NestedStruct>(&json!({
        "ts": [{"e": "bar"}, {"e": "foo"}],
    }).to_string()));
    println!("serde_qs: {:?}", serde_qs::from_str::<NestedStruct>(
        "ts[0][e]=foo&ts[1][e]=bar"
    ));
}
```

Solving this may be as easy as moving tuple_struct back to forward_to_deserialize_any somewhere in the code. AFAIK it's impossible to write a custom deserializer just for one tuple variant of an enum (you need to do it for the entire enum), so this should not interfere with the new support for unit & tuple variants.

**Comment by Xion** [comment]
Oh one more completely unrelated thing. The HEAD version has hidden the `serde_qs::de::Error` type. It'd be nice to expose as just `serde_qs::Error`.

**Comment by samscott89** [comment]
Ah okay, that's a nice point. I was thinking that I was helping implementations by forcing the "correct" visitor implementation on them, but I can see the problem here is that you are actually taking away information from the visitor.
Incidentally, that's a horrendous use case 😛. Great though! I wish there was a list of odd use cases to target support for. (Technically foo and bar are the wrong way around if you care about ordering btw).

I _think_ I still need it to coerce some values to a sequence. For example, I want newtype structs to basically just act like the inner value, which means you need to treat a single value like a sequence of length 1. But I think that is necessary since currently the deserialize macro treats newtypes like tuples of length 1. 

Thanks for the feedback. I'll try to get my head around using error chain properly to expose nice, chainable errors.

**Comment by samscott89** [comment]
Okay, how about now?

**Comment by Xion** [comment]
I'm not sure you can just [whack-a-mole the possible input types like that](https://github.com/samscott89/serde_qs/commit/6429b975c6cbcf5d7e0c2b1d02026f8d46b5c50f#diff-d95a844fba94db5afc697ebd6eda4e5aR618), because one can always add something non-standard (the `visit_str` is new):
```rust
             extern crate serde;
#[macro_use] extern crate serde_derive;
#[macro_use] extern crate serde_json;
             extern crate serde_qs;


use std::fmt;

use serde::de::{self, Deserialize, Visitor};


#[derive(Deserialize, Debug, Clone, Copy)]
#[serde(rename_all = "lowercase")]
enum TestEnum { Foo, Bar }

#[derive(Deserialize, Debug)]
struct TestStruct {
    e: TestEnum,
}


#[derive(Debug)]
struct TupleStruct(TestEnum, TestStruct);

impl<'de> Deserialize<'de> for TupleStruct {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
        where D: de::Deserializer<'de>
    {
        deserializer.deserialize_tuple_struct(
            "TupleStruct", 2, TupleStructVisitor)
    }
}

struct TupleStructVisitor;
impl<'de> Visitor<'de> for TupleStructVisitor {
    type Value = TupleStruct;

    fn expecting(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "TupleStruct representation")
    }

    fn visit_str<E: de::Error>(self, s: &str) -> Result<Self::Value, E> {
        let e = match s {
            "alt_foo" => TestEnum::Foo,
            "alt_bar" => TestEnum::Bar,
            _ => return Err(E::custom("bad!")),
        };
        let result = TupleStruct(e, TestStruct { e });
        Ok(result)
    }

    fn visit_map<V>(self, map: V) -> Result<Self::Value, V::Error>
        where V: de::MapAccess<'de>
    {
        let inner_de = de::value::MapAccessDeserializer::new(map);
        let test_struct: TestStruct = Deserialize::deserialize(inner_de)?;

        let result = TupleStruct(test_struct.e, test_struct);
        Ok(result)
    }
}


#[derive(Deserialize, Debug)]
struct NestedStruct {
    ts: Vec<TupleStruct>,
}


fn main() {
    println!("serde_json: {:?}", serde_json::from_str::<NestedStruct>(&json!({
        "ts": ["alt_bar", "alt_foo"],
    }).to_string()));
    println!("serde_qs: {:?}", serde_qs::from_str::<NestedStruct>(
        "ts[0]=alt_foo&ts[1]=alt_bar"
    ));
}
```

```
serde_json: Ok(NestedStruct { ts: [TupleStruct(Bar, TestStruct { e: Bar }), TupleStruct(Foo, TestStruct { e: Foo })] })
serde_qs: Err(Error(Custom("invalid type: sequence, expected TupleStruct representation"), State { next_error: None, backtrace: None }))
```

As for my actual case, I like to think is not as horrendous :) I'm basically deserializing a list which may be of strings or of maps/structs, but I need to remember which one was it for the outer deserializer -- hence the tuple struct with additional "source" enum tag.

The real code is [here](https://github.com/Xion/rofld/blob/4ec9d11295ce51130d3982045e1b5cde4855ec9e/src/model/de/image_macro.rs#L253) -- I'm sure you will find this hack almost as amusing as the entire project ;) (The last test case that still doesn't pass with this version is [here](https://github.com/Xion/rofld/blob/4ec9d11295ce51130d3982045e1b5cde4855ec9e/src/model/de/tests.rs#L427)).

**Comment by samscott89** [comment]
Hmm, okay, I think I get it now. Kind of tough with this format, since it's completely unspecified.

To make this work then I'll have to make it so that newtypes structs are serialized the same as tuple structs, so that the format is picked up and no tricks are needed to get the "String -> sequence -> newtype" trick to work.

On reviewing your use case, I'm not sure I can continue working on this repo in good conscience... For something this important you'd need someone with _at least_ a decade of Rust experience to get it right. It can't afford to be any other way.

**Comment by samscott89** [comment]
I've been giving it some more though, and I'm not entirely sure you _can_ expect the Deserializer to necessarily call the correct `visit_*` method.

As described [here](https://serde.rs/impl-deserialize.html#driving-a-visitor):
> Note that a Deserializer will not necessarily follow the type hint, so the call to deserialize_i32 does not necessarily mean the Deserializer will call I32Visitor::visit_i32.

Basically, you're telling the Deserializer that it should expect a TupleStruct, and then expecting it to work out that it's actually still okay to call `visit_str` or `visit_map` for it? But that's only going to work for self-describing formats.

So it works for JSON, it _could_ work in this exact case for `serde_qs`, but in general I don't think it can, and it certainly wouldn't work for something like [bincode](https://github.com/TyOverby/bincode) (for when you port your application to a compact binary format to save on bandwidth of course).

**Comment by Xion** [comment]
Sorry, didn't want to discourage you :) Remember that no one _can_ have a decade of experience in Rust yet! No software is perfect, we are all still learning here, and -- like I mentioned at the very beginning -- workarounds exist.

**Comment by Xion** [comment]
> So it works for JSON, it could work in this exact case for serde_qs, but in general I don't think it can, and it certainly wouldn't work for something like bincode (for when you port your application to a compact binary format to save on bandwidth of course).

Yes, the deserializers I wrote were specifically tailored for self-describing, tree-like structures -- most notably JSON. I was actually kind of surprised it is possible to encode those in query strings, but seeing the format `serde_qs` uses made it seem very simple.

Indeed, I think the scheme you're using *is* completely isomorphic with JSON, as it is essentially a linear list of all key paths. In a pinch, it should be possible to just parse it first to `HashMap<String, Stuff>` (or `HashMap<Index, Stuff>` to account for numeric indices) and work with that.

In any case, I think the most important part is to make it clear about the limitations of the current (or any) version. You are definitely welcome to simply add a note in README about the assumptions your deserializer makes and just close this issue :)

**Comment by samscott89** [comment]
> Indeed, I think the scheme you're using is completely isomorphic with JSON, as it is essentially a linear list of all key paths. 

Yeah, it started out as an experiment to see if it was possible, and took a lot of inspiration from the JSON code. I'm actually kind of tempted to just implement the whole thing by deserializing to `serde_json::Value` instead. The minor problem is that String aren't properly delimited in querystrings.

I'll have a think if there's a nice way to keep it flexible enough to support custom deserializers expected odd formats, but without degrading the experience for derived deserializations.

Thanks for discussing this one out :) Helps my understanding of all this, and eventually the library as well.

**Comment by Xion** [comment]
>  The minor problem is that String aren't properly delimited in querystrings.
They aren't? I've seen plenty of web frameworks that parse QS's into some equivalent of `HashMap<String, Vec<String>>`. It was actually this plus the fact that either HTTP or browsers limit the QS size (2048 bytes or so) that made me think of just parsing the whole "AST" upfront.

But of course that may be a complete overhaul of your entire approach so take it with a grain of salt :)

**Comment by samscott89** [comment]
Right, a better way to phrase it is that querystrings are informally defined as nested maps where keys and (leaf) values are both strings. What I mean is that there is no distinction between bools/numbers/units and strings, like in JSON `"null"` vs `null` or `"10.3"` vs `10.3`.

For example, is `a[0] = 1` either: `[1u8; 1]`, `["1"; 1]`, or even a `Map<u8, String>`, ... etc.

I could definitely just return this nested map of Strings instead. This would be really simple, but not very useful in the `serde` sense.

**Comment by samscott89** [comment]
I think you can make your code work in a slightly more robust way all around:
```rust
impl<'de> Deserialize<'de> for TupleStruct {

    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
        where D: de::Deserializer<'de>
    {
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum UntaggedEnum {
            E(TestStruct),
            S(String),
        }
        let us = UntaggedEnum::deserialize(deserializer)?;

        match us {
            UntaggedEnum::E(ts) => {
                Ok(TupleStruct(ts.e, ts))
            },
            UntaggedEnum::S(s) => {
                de::Deserializer::deserialize_tuple_struct(s.into_deserializer(), "TupleStruct", 2, TupleStructVisitor)
            }
        }
    }
}

struct TupleStructVisitor;
impl<'de> Visitor<'de> for TupleStructVisitor {
    type Value = TupleStruct;

    fn expecting(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "TupleStruct representation")
    }

    fn visit_str<E: de::Error>(self, s: &str) -> Result<Self::Value, E> {
        let e = match s {
            "alt_foo" => TestEnum::Foo,
            "alt_bar" => TestEnum::Bar,
            _ => return Err(E::custom("bad!")),
        };
        let result = TupleStruct(e, TestStruct { e });
        Ok(result)
    }
}
```

This works with the current version in the repo, requires less code (don't need to manually `visit_map` any more), and (to me at least) makes more sense. Semantically, you are trying to deserialize a TupleStruct from either a struct or a String, which is exactly what this code achieves. Instead of assuming the deserializer can make this distinction, using untagged enums makes it more explicit on your side and will attempt to deserialize each one.

**Comment by Xion** [comment]
Ah, it may very well be that this is in fact the correct way. I remember reading through the custom deserializers' examples in Serde docs and thinking "wait, this is doing it in a roundabout way, I can make it shorter", so it's likely I have taken some (too radical shortcuts) :)

I'll try this when I get home.

**Comment by samscott89** [comment]
I get the impression that there is often no "correct way" in serde. But this seems like a nice compromise where I can let `serde_qs` take as many hints as it can from which `deserialize_*` method is called, but you can still make sure the right visitor method is called.

**Comment by Xion** [comment]
So this seems to have worked :) There is another type that I seem to have problems with now but I'll try to solve using a similar solution. Will report soon.

**Comment by Xion** [comment]
Alright, fixed all. I applied the "lookahead enum" technique in one more spot (https://github.com/Xion/rofld/commit/dca9e0e40f15c7998dc2421cf5dcbf19b6df4eda#diff-362ff1d9014c27059ccc2de11400e306) and it works nicely. I could in theory get rid of the `Visitor` in that file but I kinda like how it organizes the code right now.

Ultimately, I think the whole problem stems from Serde's unclear deserialization protocol, and the leaky abstraction it provides. On one hand, authors of data format deserializers (like you) need to take every possible action to determine the shape of the input beforehand (so they can calls the right `deserialize_foo`). On the other hand, people writing deserialization for their types need to _provide_ the information somehow, and some techniques are more robust here than others.

So basically, it's a balance, but Serde basically leaves us to figure it out ourselves. I predict many more issues like this arising with different `serde_foo` crates when more people try to write some wonky deserialization code.

Anyway! :) I'm content with the way it turned out, so feel free to close the issue. You may want to summarize the problem & the solution in your docs, though to be fair the best place for it would be Serde docs.

Thanks again!

**Comment by dtolnay** [comment]
It seems like there is a misunderstanding about what the various Deserializer methods mean. In particular, in this code https://github.com/Xion/rofld/commit/dca9e0e40f15c7998dc2421cf5dcbf19b6df4eda#diff-362ff1d9014c27059ccc2de11400e306 neither the "before" nor "after" code makes sense to me.

In the "before" code, the `deserialize_tuple_struct` method means that the type's representation in the [Serde data model](https://serde.rs/data-model.html) is a tuple struct, i.e. the Deserializer should follow up with a call to `visit_seq`. The bug in the original code is that `visit_seq` is not implemented on the relevant Visitor.

In the "after" code, I don't understand the lookhead enum thing. It is equivalent to calling `deserializer.deserialize_any(ColorVisitor)`.

Basically if your Deserialize implementation knows the Serde data model representation of the type that the input contains, call the right Deserializer method. If the Deserializer needs to figure it out, call `deserialize_any`.

I skimmed the discussion and it sounds like `serde_qs` had the correct behavior all along since the first refactor in https://github.com/samscott89/serde_qs/issues/6#issuecomment-302934235, and the user's Deserialize impl was badly behaved.

**Comment by Xion** [comment]
> In the "before" code, the deserialize_tuple_struct method means that the type's representation in the Serde data model is a tuple struct, i.e. the Deserializer should follow up with a call to visit_seq. The bug in the original code is that visit_seq is not implemented on the relevant Visitor.

I see know -- I have misunderstood the purpose of `Deserializer::deserialize_foo` methods. I thought you are supposed to invoke them when your final data is of shape `foo` (so for tuple structs, it'd be `deserialize_tuple_struct`). But now I understand it's about what you _expect in the input_ -- which may have only tenuous relation to the final type being deserialized.

So I guess apologies to @samscott89 are in order :) The original enum bug was a legitimate issue, but the rest turned out to be trying to whip `serde_qs` into supporting ill-behaved `Deserialize` impls which was likely unnecessary. I feel like I should atone by writing a blog post about all this `Deserialize`, `Deserializer`, and `Visitor` stuff so others won't repeat my mistakes :)

This saying, the remarks about Serde not doing a very good job at documenting such pitfalls still stand ;p

**Comment by samscott89** [comment]
Hey @dtolnay! Thanks for popping by to impart some wisdom. I was actually planning on writing this up a little better before asking for your opinion on the whole thing.
Since you already found us though...

> In the "after" code, I don't understand the lookhead enum thing. It is equivalent to calling deserializer.deserialize_any(ColorVisitor).

> Basically if your Deserialize implementation knows the Serde data model representation of the type that the input contains, call the right Deserializer method. If the Deserializer needs to figure it out, call deserialize_any.

That's a nice summary which would have made things a bit clearer earlier (still, many thanks to @Xion, I think we both learnt a lot through this exchange).

One thing I'm not too sure about still: if the `Deserializer` kind of knows what format the data is in, should it still make an effort to coerce it into the right/expected type? For example, `serde_json` wont try to convert a map into a sequence (doing something like `.map(|(k, v)| v)` perhaps) even if `deserialize_seq` is called, since JSON is self-describing so a map would be invalid. On the other hand, with these slightly informally-defined querystrings, you need a bit of help from the deserializer, so is it a reasonable thing to do to convert to the expected format?

**Comment by dtolnay** [comment]
I would recommend against preemptively implementing those coercions but if somebody asks for a specific one and the coercion is reasonable, it is fine to implement. Basically that falls under the Serializer and Deserializer's right to define the mapping between Serde data model <-> the format. That mapping is an art rather than a science. For example serde_json's mapping of an enum like `E::V { x: 0 }` <-> `{"V":{"x":0}}` is one of many reasonable choices for that mapping. Haskell and Java libraries typically map the equivalent data structure in a different way.

**Comment by samscott89** [comment]
Okay, thanks!
My only real issue is the ambiguity between a `Map<{integer}, Thing>`, and a sequence `Vec<Thing>`, since something like `a[0]=blah` could easily be parsed as either. Hence needing the extra hints.
I think I could avoid most of these issues if integer keys were restricted for array indices and never as keys for a map, and then the key map/seq/value distinction is a bit more obvious.

**Comment by Xion** [comment]
Ideally, `a` would be a sequence if all keys can be parsed as numbers. Though that of course precludes incremental deserializing of individual elements, so it's a tradeoff...

**Comment by samscott89** [comment]
Okay, so 0.2.0 is pushed and now detects sequences properly, which means that less coercing between types is needed.
Enums are definitely supported at this point, which solves the original issue 😄 .

Thanks both for your help on this.
